src/appleseed/foundation/meta/tests/test_imagefilewriter.cpp

```diff
@@ -0,0 +1,200 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// appleseed.foundation headers.
+#include "foundation/image/color.h"
+#include "foundation/image/colorspace.h"
+#include "foundation/image/genericimagefilereader.h"
+#include "foundation/image/genericimagefilewriter.h"
+#include "foundation/image/image.h"
+#include "foundation/image/imageattributes.h"
+#include "foundation/image/pixel.h"
+#include "foundation/math/scalar.h"
+#include "foundation/utility/iostreamop.h"
+#include "foundation/utility/test.h"
+
+// Standard headers.
+#include <cassert>
+#include <cmath>
+#include <cstddef>
+#include <cstring>
+#include <exception>
+#include <memory>
+
+using namespace foundation;
+
+TEST_SUITE(Foundation_Image_ImageFileWriter)
+{
+    template <typename Color>
+    void write_image(const Color fill_color, const PixelFormat pixel_format, const std::size_t image_size, const std::string image_file_path)
+    {
+        std::unique_ptr<Image> image(new Image(image_size, image_size, image_size, image_size, fill_color.Components, pixel_format));
+        image->clear(fill_color);
+        
+        GenericImageFileWriter writer(image_file_path.c_str());
+        writer.append_image(image.get());
+        writer.write();
+    }
+    
+    TEST_CASE(WriteBMP_CorrectlyWritesImagePixels)
+    {
+        const std::string image_file_path = "unit tests/outputs/test_imagefilewriter.bmp";
+        const std::size_t image_size = 16;
+        const Color4b fill_color(50, 100, 150, 42);
+        
+        write_image<Color4b>(fill_color, PixelFormatUInt8, image_size, image_file_path);
+
+        GenericImageFileReader reader;
+        std::unique_ptr<Image> image(reader.read(image_file_path.c_str()));
+            
+        for (std::size_t y = 0; y < image_size; ++y)
+        {
+            for (std::size_t x = 0; x < image_size; ++x)
+            {
+                Color4b c;
+                image->get_pixel(x, y, c);
+                EXPECT_EQ(fill_color, c);
+            }
+        }
+    }
+    
+    TEST_CASE(WriteEXR_CorrectlyWritesImagePixels)
+    {
+        const std::string image_file_path = "unit tests/outputs/test_imagefilewriter.exr";
+        const std::size_t image_size = 16;
+        const Color4b fill_color(50, 100, 150, 42);
```

Comment:
For the OpenEXR format it would make sense to read and write floating point values since that's what the format natively supports.